### PR TITLE
Updates to modules for H2C

### PR DESCRIPTION
**Pull request by nkern**
This PR makes the `abscal` and `lstbin` module compatible with `HERAData` objects

Implementation of partial loading in `abscal` and `lstbin` algorithms and Py3 compatibility are deferred to a future PR

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -26,7 +28,7 @@ install:
   - source activate test-environment
   - conda install -c conda-forge aipy pycodestyle
   - pip install coveralls
-  - pip install git+https://github.com/HERA-Team/pyuvdata.git@v1.3
+  - pip install git+https://github.com/HERA-Team/pyuvdata.git@9c94be0366eb41692f5501caac99981026a7ffe6
```
Comment (jsdillon) [inline .travis.yml]: Can't we get up to date with master? I'd like to try to always be up to date with master and only build against a specific version or commit when master breaks us.

File: hera_cal/abscal.py [hunk 2]
```diff
@@ -83,34 +81,38 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
 
         Parameters:
         -----------
-        model : visibility data of refence model, type=DataContainer
+        model : Visibility data of refence model, type=dictionary
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Why this change? Are we never using any of DataContainer's special properties?

File: hera_cal/abscal.py [hunk 3]
```diff
@@ -83,34 +81,38 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
 
-        wgts : weights of data, type=DataContainer, [default=None]
+        wgts : weights of the data, type=dictionary, [default=None]
                keys are antenna pair + pol tuples (must match model), values are real floats
                matching shape of model and data
 
         antpos : type=dictionary, dict of antenna position vectors in ENU (topo) frame in meters.
                  origin of coordinates does not matter, but preferably are centered in the array.
-                 keys are antenna integers and values are 2D ndarray position vectors,
+                 keys are antenna integers and values are 2D position East-North arrays,
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Can't they also be East-North-Up with Up in position [2]?

File: hera_cal/abscal.py [hunk 4]
```diff
@@ -122,25 +124,28 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
 
         bl_taper_fwhm : float, impose a gaussian taper on the data weights as a function of
             bl separation length, with a specified fwhm [meters]
         """
         # set pols to None
         pols = None
 
-        # check format of model if model is not a dictionary
+        # load model if necessary
         if isinstance(model, list) or isinstance(model, np.ndarray) or isinstance(model, str) or isinstance(model, UVData):
```
Comment (jsdillon) [inline hera_cal/abscal.py]: You might want to use `issubclass(UVData)` so `HERAData` works too.

File: hera_cal/abscal.py [hunk 5]
```diff
@@ -122,25 +124,28 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
             (model, model_flags, model_antpos, model_ants, model_freqs, model_lsts,
              model_times, model_pols) = io.load_vis(model, pop_autos=True, return_meta=True)
 
-        # check format of data if data is not a dictionary
+        # load data if necessary
         if isinstance(data, list) or isinstance(data, np.ndarray) or isinstance(data, str) or isinstance(data, UVData):
```
Comment (jsdillon) [inline hera_cal/abscal.py]: And likewise here.

File: hera_cal/abscal.py [hunk 6]
```diff
@@ -83,34 +81,38 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
-                Optionally, model can be a path to a miriad or uvfits file, or a
-                pyuvdata.UVData object, or a list of either.
+                Optionally, model can be a path to a miriad file, a
+                pyuvdata.UVData object or hera_cal.HERAData object,
+                or a list of either.
 
-        data : visibility data of measurements, type=DataContainer
-               keys are antenna pair + pol tuples (must match model), values are
-               complex ndarray visibilities matching shape of model
+        data :  Visibility data, type=dictionary
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Same comment here.

File: hera_cal/abscal.py [hunk 7]
```diff
@@ -83,34 +81,38 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
+                pyuvdata.UVData object or hera_cal.HERAData object,
+                or a list of either. In this case, antpos, freqs
+                and wgts are overwritten from arrays in data.
 
         refant : antenna number integer for reference antenna
             The refence antenna is used in the phase solvers, where an absolute phase is applied to all
             antennas such that the refant's phase is set to identically zero.
 
-        wgts : weights of data, type=DataContainer, [default=None]
+        wgts : weights of the data, type=dictionary, [default=None]
```
Comment (jsdillon) [inline hera_cal/abscal.py]: And here.

File: hera_cal/abscal.py [hunk 8]
```diff
@@ -1457,15 +1459,22 @@ def cut_bls(datacontainer, bls, bl_cut):
...
-    Output (cut_datacontainers)
+    Output
     ------
     cut_datacontainer : DataContainer object with bl cut enacted
     """
+    if min_bl_cut is None:
+        min_bl_cut = 0.0
+    if max_bl_cut is None:
+        max_bl_cut = 1e10
     cut_datacontainer = odict()
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Because data containers now carry around some metadata, I think this is a bad solution. Instead, I'd make a deepcopy and use `del`.

File: hera_cal/abscal.py [hunk 9]
```diff
@@ -1446,9 +1448,9 @@ def abscal_run(data_file, model_files, refant=None, calfits_infile=None, verbose
         return return_obj
 
 
-def cut_bls(datacontainer, bls, bl_cut):
+def cut_bls(datacontainer, bls, min_bl_cut=None, max_bl_cut=None):
```
Comment (jsdillon) [inline hera_cal/abscal.py]: It seems to me like `bls` should be an optional parameter, defaulted to None, an `if bls is None: bls = datacontainer.keys()`

File: hera_cal/abscal_funcs.py [hunk 10]
```diff
@@ -1676,6 +1667,7 @@ def avg_file_across_red_bls(data_fname, outdir=None, output_fname=None,
 
     # write to file
     if write_miriad:
+        raise NotImplementedError("Correct averaging of integration_time and nsample not yet implemented")
```
Comment (jsdillon) [inline hera_cal/abscal_funcs.py]: This is a bit strange. You definitely want to comment out the code that isn't usable. But I wonder: why aren't we using `io.write_vis`? Can't we just eliminate this whole block?

File: hera_cal/abscal.py (hunk 2 above)
Comment (nkern) [reply hera_cal/abscal.py]: I thought you wanted to deprecate `DataContainer`, so I thought it would just be apt to use a dictionary, which I think is actually the only thing we need, but I can add `or DataContainer` if you planned to keep it around...

File: hera_cal/abscal_funcs.py (hunk 10 above)
Comment (nkern) [reply hera_cal/abscal_funcs.py]: I punted on writing b/c we need to implement the proper summing of the `nsample` and `integration_time` arrays which is not trivial and I didn't want to invest time into it now. I can comment out that code if you like

File: hera_cal/abscal.py (hunk 2 above)
Comment (jsdillon) [reply hera_cal/abscal.py]: No, the whole point of `HERAData` is that it spits out DataContainers. We're definitely keeping them.

File: hera_cal/abscal_funcs.py (hunk 10 above)
Comment (jsdillon) [reply hera_cal/abscal_funcs.py]: Yeah, I'd comment out the code. Otherwise our coverage will go way down.

Alternatively, if no one is using the code, we could also delete it.

File: hera_cal/abscal.py (hunk 8 above)
Comment (jsdillon) [reply hera_cal/abscal.py]: Also, there may be other places where you turn dicts/odicts into DataContainers. We should avoid that if possible.

**Comment by jsdillon** [comment]
@nkern I'd really like to get the pyuvdata update done before everything else. Can we narrow the scope of this PR and do the rest in a different PR?

File: hera_cal/abscal.py (hunk 9 above)
Comment (nkern) [reply hera_cal/abscal.py]: bls is not a list of keys, its a dictionary of baseline vectors

File: hera_cal/abscal.py [hunk 11]
```diff
@@ -83,28 +81,32 @@ def __init__(self, model, data, refant=None, wgts=None, antpos=None, freqs=None,
...
+        data :  Visibility data, type=dictionary or DataContainer
+                keys are antenna-pair + polarization tuples, Ex. (1, 2, 'xx').
+                values are complex ndarray visibilities.
+                these must be 2D arrays, with [0] axis indexing time
+                and [1] axis indexing frequency.
 
-                Optionally, model can be a path to a miriad or uvfits file, or a
-                pyuvdata.UVData object, or a list of either. In this case, wgts, antpos,
-                freqs, times and pols are overwritten with equivalent information from data object.
+                Optionally, data can be a path to a miriad file, a
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Do we want to support uvh5 rather than miriad? Or just leave this as a legacy interface?

File: hera_cal/abscal.py [hunk 12]
```diff
@@ -1446,28 +1450,42 @@ def abscal_run(data_file, model_files, refant=None, calfits_infile=None, verbose
...
+    max_bl_cut : float, maximum baseline separation [meters] to keep in data    
 
-    bl_cut : float, maximum baseline separation [meters] to keep in data
+    inplace : bool, if True edit data in input object, else make a copy.
 
-    Output (cut_datacontainers)
+    Output
     ------
-    cut_datacontainer : DataContainer object with bl cut enacted
+    datacontainer : DataContainer object with bl cut enacted
```
Comment (jsdillon) [inline hera_cal/abscal.py]: I think it's generally better to keep the returns of the same form whenever possible... so in this case, I'd recommend returning the cut datacontainer regardless. If people want to use inplace, it's OK to have a return and just call it without any assignment operator.

File: hera_cal/abscal.py [hunk 13]
```diff
@@ -1446,28 +1450,42 @@ def abscal_run(data_file, model_files, refant=None, calfits_infile=None, verbose
...
+        only returned if not inplace
     """
-    cut_datacontainer = odict()
+    if inplace:
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Shouldn't this be `if not inplace`?

File: hera_cal/abscal.py [hunk 14]
```diff
@@ -1446,28 +1450,42 @@ def abscal_run(data_file, model_files, refant=None, calfits_infile=None, verbose
         return return_obj
 
 
-def cut_bls(datacontainer, bls, bl_cut):
+def cut_bls(datacontainer, bls, min_bl_cut=None, max_bl_cut=None, inplace=False):
```
Comment (jsdillon) [inline hera_cal/abscal.py]: Data containers (if they come from a file) now contain `antpos` inside them. Do we want to remove the dependence on `bls`?

File: hera_cal/abscal.py [hunk 15]
```diff
@@ -1446,28 +1450,42 @@ def abscal_run(data_file, model_files, refant=None, calfits_infile=None, verbose
...
+        if k not in bls:
+            continue
+        if bl_len > max_bl_cut or bl_len < min_bl_cut:
+            del datacontainer[k]
 
-    assert len(cut_datacontainer) > 0, "no baselines were kept after baseline cut..."
+    assert len(datacontainer) > 0, "no baselines were kept after baseline cut..."
 
-    return DataContainer(cut_datacontainer)
+    if not inplace:
```
Comment (jsdillon) [inline hera_cal/abscal.py]: See above comment.

File: hera_cal/io.py [hunk 16]
```diff
@@ -700,10 +697,11 @@ def write_vis(fname, data, lst_array, freq_array, antpos, time_array=None, flags
...
     telescope_location : type=ndarray, telescope location in xyz in ITRF (earth-centered frame).
 
-    integration_time : type=float or ndarray, integration duration in seconds for data_array. This does not necessarily have
-        to be equal to the diff(time_array): for the case of LST-binning, this is not the duration of the LST-bin
-        but the integration time of the pre-binned data. If None, will use diff(time_array) by default.
-        
+    integration_time : type=float or ndarray, integration duration in seconds for data_array.
+        This does not necessarily have to be equal to the diff(time_array): for the case of
+        LST-binning, this is not the duration of the LST-bin but the integration time of the
+        pre-binned data. Default is median(diff(time_array)) in seconds.
```
Comment (jsdillon) [inline hera_cal/io.py]: I would clarify here that the total integration time is always `integration_time * nsamples`.

File: hera_cal/abscal.py (hunk 11 above)
Comment (nkern) [reply hera_cal/abscal.py]: I added a filetype kwarg to support both

File: hera_cal/abscal.py (hunk 13 above)
Comment (nkern) [reply hera_cal/abscal.py]: yup, good catch thanks

File: hera_cal/abscal.py (hunk 12 above)
Comment (nkern) [reply hera_cal/abscal.py]: sure thing

File: hera_cal/io.py (hunk 16 above)
Comment (nkern) [reply hera_cal/io.py]: sure thing

File: hera_cal/abscal.py (hunk 11 above)
Comment (jsdillon) [reply hera_cal/abscal.py]: Where? I don't see it.

File: hera_cal/abscal.py (hunk 14 above)
Comment (jsdillon) [reply hera_cal/abscal.py]: No?

Comment (nkern) [reply hera_cal/abscal.py]: ok
